portfolio/src/main/java/com/google/sps/servlets/ThumbsDownServlet.java

```diff
@@ -31,44 +31,82 @@
 import com.google.appengine.api.datastore.Query.SortDirection;
 import com.google.appengine.api.datastore.Key;
 import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
 
 @WebServlet("/thumbsdown-data")
 public final class ThumbsDownServlet extends HttpServlet {
 
     @Override
-    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {        
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        UserService userService = UserServiceFactory.getUserService();
+        if (!userService.isUserLoggedIn()) {
+            response.sendRedirect("/contact.html");
+            return;
+        }
+
+        Entity userInfoEntity = getUserInfoEntity();
+
         // Get comment's id (which was passed as a parameter).
         long id = Long.parseLong(request.getParameter("id"));
 
+        Entity commentEntity = getCommentEntity(id);
+        if (commentEntity == null) {
+            response.setContentType("text/html;");
+            response.getWriter().println("Unable to get comment.");
+            return;
+        }
+        
+        if (!unlikedComment(userInfoEntity, commentEntity)) {
```

Comment:
```suggestion
        if (!isUnlikedComment(userInfoEntity, commentEntity)) {
```

Reply:
It might be clearer if the method name is a verb phrase.

Reply:
I agree. Just added this one